### DISCUSSION TITLE
Removing `Validation` from phases that still use `HardFailure`

### DISCUSSION THREAD

**Discussion by jaschdoc** [flix/flix#9025] [Ideas]
Goal: we want to remove the `Validation` dependency so we can drop the monadic way of plumbing phases together.
Problem: we still use `HardFailure` in some phases since we have not yet made those errors recoverable.

We can *temporarily* use exceptions instead of `HardFailure` in `Typer` and `Resolver` and catch them within the phase (so we can collect all hard failures). That way the errors "bubble up" to a place where we can add them to the shared context (in a new field) while preserving the semantics of the phase, i.e., the phase stops its computations for that failure.
We then add (maybe we generalize to `A, B, C`)
```scala
def phaseUnrecoverable(phaseName: String, phase: => (Root, CompilationMessage & Recoverable, CompilationMessage & Unrecoverable): (Root, CompilationMessage & Recoverable, CompilationMessage & Unrecoverable) = ???
```
As mentioned before, we can drop the for-comprehension and do some special casing after `Typer` and `Resolver` to check that the list of unrecoverable errors is empty or abort the rest of the pipeline otherwise.

Exceptions may be slightly evil but it helps us get rid of `Validation` sooner which currently makes the code slightly awkward as most phases do not use it anymore.

**Comment by jaschdoc** [comment]
Also, we are close to being free from `HardFailure`s in `Weeder2` but if we do not come up with an elegant solution for `pick` https://github.com/flix/flix/pull/8824 then we can use this trick here too. This will let us completely remove `Validation` from the phase then.

**Comment by magnus-madsen** [comment]
We can discuss this, but right now the main concern is to get rid of SoftFailure.
